test/ion/oracles/GenericRateProvider.t.sol

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.21;
+
+import { GenericRateProvider } from "./../../../src/helper/GenericRateProvider.sol";
+import { SWBTC } from "./../../../src/helper/Constants.sol";
+
+import "forge-std/Test.sol";
+
+abstract contract GenericRateProviderTest is Test {
+    GenericRateProvider rateProvider;
+    address target;
+    bytes4 selector;
+    bytes32 staticArgument0;
+    bytes32 staticArgument1;
+    bytes32 staticArgument2;
+    bytes32 staticArgument3;
+
+    uint8 decimals;
+
+    function setUp() public virtual {
+        vm.createSelectFork(vm.envString(_getRpcUrl()));
+
+        _initialize();
+
+        require(address(rateProvider) != address(0), "rate provider not set");
+        require(decimals > 0, "decimals not set");
+        require(target != address(0), "target not set");
+        require(bytes4(selector).length > 0, "selector not set");
+    }
+
+    function test_GetRateWithinExpectedBounds() public {
+        uint256 rate = rateProvider.getRate();
+        (uint256 min, uint256 max) = _expectedRateMinMax();
+
+        require(rate >= min, "rate must be greater than or equal to min");
+        require(rate <= max, "rate must be less than or equal to max");
+    }
+
+    function _initialize() public virtual;
+
+    function _expectedRateMinMax() public virtual returns (uint256, uint256);
+
+    function _getRpcUrl() public pure virtual returns (string memory);
+}
+
+contract SwBtcRateProviderTest is GenericRateProviderTest {
+    function _initialize() public override {
+        target = SWBTC;
+        selector = bytes4(keccak256("pricePerShare()"));
+        decimals = 8;
+        staticArgument0 = bytes32(uint256(123));
```

Comment:
is this just to prove that having a non-zero static argument doesn't interfere with getRate?

Reply:
Yes just making sure that it gets ignored for this call since the function does not take in any other params